sextant/src/npk.rs

```diff
@@ -286,72 +286,86 @@ fn gen_pseudo_files(manifest: &Manifest) -> Result<Vec<(&str, u32)>> {
 fn gen_hash_tree(
     image: &File,
     image_size: u64,
-    block_size: u64,
     salt: &[u8; SHA256_SIZE],
-    hash_level_offsets: &[usize],
+    level_offsets: &[usize],
     tree_size: usize,
 ) -> Result<(Vec<u8>, Vec<u8>)> {
+    // For a description of the overall hash tree generation logic see
+    // https://source.android.com/security/verifiedboot/dm-verity#hash-tree
+
     let mut hash_tree = vec![0_u8; tree_size];
-    let hash_src_offset = 0;
-    let mut hash_src_size = image_size;
     let mut level_num = 0;
-    let mut reader = BufReader::new(image);
-    let mut level_output: Vec<u8> = vec![];
+    let mut level_size = image_size;
+    let mut level_hashes: Vec<[u8; SHA256_SIZE]> = vec![];
+
+    if image_size % BLOCK_SIZE as u64 != 0 {
+        return Err(anyhow!(
+            "Failed to generate verity has tree. The image size {} is not a multiple of the block size {}",
+            image_size,
+            BLOCK_SIZE
+        ));
+    }
+
+    loop {
+        level_hashes.clear();
```

Comment:
The loop logic is not clear to me...I mean the exit condition. Breaking from a loop is always harder to read IMHO

Reply:
This attempts to stay close to the original algorithm description (see above). I'll see if this can be simplified.